src/sql/parts/notebook/notebook.component.ts

```diff
@@ -331,6 +337,47 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		}
 	}
 
+	// We can keep this method in common utilities
+	private getFilePathFromRecentWorkspace(untitledResource: URI): string {
+		let untitledFileName = untitledResource.path + '.' + DEFAULT_NOTEBOOK_FILETYPE;
+
+		let lastActiveFile = this.historyService.getLastActiveFile();
+		if (lastActiveFile) {
+			return URI.file(paths.join(paths.dirname(lastActiveFile.fsPath), untitledFileName)).fsPath;
+		}
+
+		let lastActiveFolder = this.historyService.getLastActiveWorkspaceRoot('file');
+		if (lastActiveFolder) {
+			return URI.file(paths.join(lastActiveFolder.fsPath, untitledFileName)).fsPath;
+		}
+		return untitledFileName;
+	}
+
+	promptForPath(defaultPath: string): TPromise<string> {
+		return this.windowService.showSaveDialog({ defaultPath });
+	}
+
+	// Entry point to save notebook
+	public async saveNotebook(): Promise<boolean> {
+		let self = this;
+		let notebookUri = this.notebookParams.notebookUri;
+		if (notebookUri.scheme === Schemas.untitled) {
+			let dialogPath = this.getFilePathFromRecentWorkspace(notebookUri);
+			return this.promptForPath(dialogPath).then(path => {
+				if (path) {
+					let newNotebookUri = URI.parse(path);
+					self._model.notebookUri = newNotebookUri;
+					self.notebookParams.notebookUri = newNotebookUri;
+					return this.save();
+				}
+				return false; // User clicks cancel
+			});
+		}
+		else {
+			return await this.save();
+		}
+	}
+
 	public async save(): Promise<boolean> {
```

Comment:
Please put this function in the public `save` method instead. That's what is defined in the INotebookEditor API and what's expected to be called.